### PR TITLE
[TM] Verb Queue - Port of TG SSverb subsystem

### PR DESCRIPTION
**Pull request by Vi3trice**
## What Does This PR Do
Ports most of https://github.com/tgstation/tgstation/pull/65589.

This doesn't port the changes made to `SSinput` nor `SSspeech`, so this system only handles verbs like `examinate`, `resist`, `quick_equip`, `mode`, `pointed` and `/client/Topic`.

A version with `/client/Click` being part of it was made and worked, but I wasn't too sure about actually queueing player clicks, so it's shelved for now.

## Why It's Good For The Game
See `TICK_ORDER.md`.

## Images of changes
Verbs

## Testing
Ran an instance, changed the `SSverb` var `always_queue` to force a verb pile up. They do, but you barely see it on a single-player instance running on a W-2135.

## Changelog
N/A

### CODE REVIEW

File: code/__DEFINES/MC.dm [hunk 1]
```diff
@@ -77,3 +86,11 @@
...
 }\
 /datum/controller/subsystem/processing/##X
+
+#define VERB_MANAGER_SUBSYSTEM_DEF(X) GLOBAL_REAL(SS##X, /datum/controller/subsystem/verb_manager/##X);\
+/datum/controller/subsystem/verb_manager/##X/New(){\
+	NEW_SS_GLOBAL(SS##X);\
+	PreInit();\
+}\
+/datum/controller/subsystem/verb_manager/##X/fire() {..() /*just so it shows up on the profiler*/} \
+/datum/controller/subsystem/verb_manager/##X
```
Comment (AffectedArc07) [inline code/__DEFINES/MC.dm:96]: Do we need this def? 

We dont use it in the file.

**Review by AffectedArc07 — COMMENTED** [review]
My only concern with this is that of "what happens to verbs if the MC is stalled or redlined?"

Do you think you can make a subsystem that will forcefully halt the MC (just put a sleep(1000) in fire) and try and execute verbs? Same with setting `processing = FALSE` on the MC.

Verbs always need to work even if the MC is fubar, since VV is a verb, same with ban panel, and those must always function regardless of the MC state.

**Comment by Vi3trice** [comment]
> My only concern with this is that of "what happens to verbs if the MC is stalled or redlined?"

This subsystem doesn't queue all verbs, only the ones that are set to be sent through it. `examinate`, `resist`, `quick_equip`, `mode`, `pointed` and `/client/Topic` are the only verbs in this PR that will be sent to a queue if the server is backed up through the verb manager defines (e.g. DEFAULT_TRY_QUEUE_VERB).

If the MC goes belly up, there is a var for it.

```
///if this is true all verbs immediately execute and dont queue. in case the mc is fucked or something
var/FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs = FALSE
```

**Comment by AffectedArc07** [comment]
> ///if this is true all verbs immediately execute and dont queue. in case the mc is fucked or something
> var/FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs = FALSE

Can you make this a verb under the `Debug` tab and call it `Enable forced verb execution` with a warning like `This will make all verbs bypass the queueing system, creating more lag. Are you absolutely sure?` 

VV relies on `/client/Topic` to change stuff, so if thats being held by a crashed SS we aint gonna be able to VV it.

**Comment by Vi3trice** [comment]
Added a verb

File: code/controllers/subsystem/verb_manager.dm [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		executed_verbs++
+
+	verb_queue.Cut()
+	verbs_executed_per_second = MC_AVG_SECONDS(verbs_executed_per_second, executed_verbs, wait SECONDS)
+	//note that wait SECONDS is incorrect if this is called outside of fire() but because byond is garbage i need to add a timer to rustg to find a valid solution
+
+/datum/controller/subsystem/verb_manager/get_stat_details()
+	return "V/S: [round(verbs_executed_per_second, 0.01)]"
+
+/client/proc/force_verb_bypass()
```
Comment (S34NW) [inline code/controllers/subsystem/verb_manager.dm:162]: shouldn't this be a verb?

Comment (AffectedArc07) [reply code/controllers/subsystem/verb_manager.dm:162]: No.

/client/verb is automatically scoped to everyone. /client/proc has to be manually added

Comment (S34NW) [reply code/controllers/subsystem/verb_manager.dm:162]: smh
image

Comment (AffectedArc07) [reply code/controllers/subsystem/verb_manager.dm:162]: > smh image

I meant in terms of interaction not code, dingus

File: code/modules/point/point.dm [hunk 3]
```diff
@@ -104,6 +104,14 @@
 	if(istype(A, /obj/effect/temp_visual/point) || istype(A, /atom/movable/emissive_blocker))
 		return FALSE
 
+	DEFAULT_QUEUE_OR_CALL_VERB(VERB_CALLBACK(src, PROC_REF(_pointed), A))
+
+/// possibly delayed verb that finishes the pointing process starting in [/mob/verb/pointed()].
+/// either called immediately or in the tick after pointed() was called, as per the [DEFAULT_QUEUE_OR_CALL_VERB()] macro
+/mob/proc/_pointed(atom/A)
```
Comment (farie82) [inline code/modules/point/point.dm:111]: Is there a reason this one is called `_pointed` while the others are formatted as `execute_XXX`?
Edit just noticed some are called `run_XXX`

File: code/__DEFINES/verb_manager.dm [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+///attempt to queue this verb process if the server is overloaded. evaluates to FALSE if queuing isnt necessary or if it failed.
+///_verification_args... are only necessary if the verb_manager subsystem youre using checks them in can_queue_verb()
+///if you put anything in _verification_args that ISNT explicitely put in the can_queue_verb() override of the subsystem youre using,
+///it will runtime.
+#define TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) (_queue_verb(_verb_callback, _tick_check, _subsystem_to_use, _verification_args))
+///queue wrapper for TRY_QUEUE_VERB() when you want to call the proc if the server isnt overloaded enough to queue
+#define QUEUE_OR_CALL_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) \
+	if(!TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args)) {\
+		_verb_callback:InvokeAsync() \
```
[the remark this answers is not in the document]
Comment (Vi3trice) [reply code/__DEFINES/verb_manager.dm:27]: Byond does, but not linters, which expect `.` to be more strict.

[the remark this answers is not in the document]
Comment (farie82) [reply code/__DEFINES/verb_manager.dm:27]: Ah yep I see, it is due to the code instantly passing the object. Not a variable with a type

**Review by S34NW — COMMENTED** [review]
looks okay? as far as I can see...

We need that TM to confirm

File: code/modules/point/point.dm [hunk 5]
```diff
@@ -104,6 +104,14 @@
 	if(istype(A, /obj/effect/temp_visual/point) || istype(A, /atom/movable/emissive_blocker))
 		return FALSE
 
+	DEFAULT_QUEUE_OR_CALL_VERB(VERB_CALLBACK(src, PROC_REF(run_pointed), A))
+
+/// possibly delayed verb that finishes the pointing process starting in [/mob/verb/pointed()].
+/// either called immediately or in the tick after pointed() was called, as per the [DEFAULT_QUEUE_OR_CALL_VERB()] macro
+/mob/proc/run_pointed(atom/A)
+	if(client && !(A in view(client.view, src)))
```
Comment (farie82) [inline code/modules/point/point.dm:112]: This will fix #13772 won't it?

Comment (Vi3trice) [reply code/modules/point/point.dm:112]: You already can't do that on master.

**Review by farie82 — APPROVED** [review]

**Pull request merged**